rfcs/NNNN-decorator-roadmap.md

```diff
@@ -0,0 +1,278 @@
+---
+Status: Accepted
+Champions: "@justinfagnani"
+PR: https://github.com/lit/rfcs/pull/20
+---
+
+# Decorator Roadmap
+
+This RFC describes how Lit can migrate to standard decorators with the smoothest upgrade path possible for developers.
+
+## Objective
+
+Migrate to using standard decorators as the only decorator implementation and unify Lit's syntax around decorators as the only way to declare reactive properties. When the migration is complete, Lit will require a decorator-supporting environment or toolchain.
+
+### Goals
+
+- Make Lit decorators work without transpilation
+- Make the migration as easy as possible on developers
+- Have a single implementation for decorators
+- Unify Lit's reactive property syntax
+- Simplify ReactiveElement's property handling code
+- Increase component performance due to more static class initialization
+- Decrease code size with simpler decorator implementations
+
+### Non-Goals
+
+- Change our decorators API drastically (ie, split `@property()` into `@input()`, `@output()`, `@attribute`, etc)
+
+## Motivation
+
+As [standard decorators](https://github.com/tc39/proposal-decorators) are starting to ship in [compilers](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-0.html#decorators) and soon in VMs, we need to prepare for migrating Lit to use them.
+
+Standard decorators will allow us to unify our surface syntax which is currently different in plain JS and compiled sources using TypeScript or Babel. This will let us use decorators as the one way of declaring reactive properties, instead of also offering the `static properties` feature. Removing `static properties` in turn lets us remove the infrastructure for dynamically creating reactive properties, simplifying ReactiveElement.
+
+## Detailed Design
+
+Detailed design of the new decorators should be covered in another RFC. This RFC focuses on:
```

Comment:
If we're going to have a separate design doc, maybe it belongs there but I don't want us to lose context on the situation described and references to the TS issues present here https://github.com/lit/lit/pull/4183#issuecomment-1711959635

Is there a place for that kind of detail here?

Reply:
Took a look at a Rust RFC for inspiration https://github.com/rust-lang/rfcs/blob/master/text/0221-panic.md and I think it's totally fine to add some sentences here that link to that PR and comment. Especially so the context doesn't get lost. And this RFC is much more findable than the comment in that PR.

Reply:
Decided to address this comment in the source code next to the relevant logic: https://github.com/lit/lit/pull/4336